bus-mapping/src/exec_trace.rs

```diff
@@ -0,0 +1,213 @@
+pub(crate) mod exec_step;
+use crate::evm::EvmWord;
+use crate::operation::{container::OperationContainer, Operation};
+use core::ops::{Index, IndexMut};
+pub(crate) use exec_step::ExecutionStep;
+use pasta_curves::arithmetic::FieldExt;
+
+/// Doc
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct BlockConstants<F: FieldExt> {
+    hash: EvmWord, // Until we know how to deal with it
+    coinbase: F,
+    timestamp: F,
+    number: F,
+    difficulty: F,
+    gaslimit: F,
+    chain_id: F,
+}
+
+/// Doc
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct ExecutionTrace<F: FieldExt> {
+    entries: Vec<ExecutionStep>,
+    block_ctants: BlockConstants<F>,
+    container: OperationContainer,
+}
+
+impl<F: FieldExt> Index<usize> for ExecutionTrace<F> {
+    type Output = ExecutionStep;
+    fn index(&self, index: usize) -> &Self::Output {
+        &self.entries[index]
+    }
+}
+
+impl<F: FieldExt> IndexMut<usize> for ExecutionTrace<F> {
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        &mut self.entries[index]
+    }
+}
+
+impl<F: FieldExt> ExecutionTrace<F> {
+    fn container_mut(&mut self) -> &mut OperationContainer {
+        &mut self.container
+    }
+
+    // Generates the trace setting the propper GC's and BusMappings for each step.
+    pub fn new(mut entries: Vec<ExecutionStep>, block_ctants: BlockConstants<F>) -> Self {
+        let mut container = OperationContainer::new();
+        let mut gc_counter = 0usize;
+
+        entries.iter_mut().for_each(|exec_step| {
+            exec_step.set_gc(gc_counter);
+            gc_counter += exec_step.gen_associated_ops::<F>(&mut container);
+            // Sum 1 to counter so that we set the next exec_step GC to the correct index
+            gc_counter += 1;
```

Comment:
Why the `ExecutionStep` itself need to increase `gc` (or say occupy a `gc`)? I was thinking it's only a wrapper of several bus mapping operations, and this is only the offset for `gen_associated_ops` to know current `gc`.

Reply:
Note that this struct is `ExecutionTrace`. And the trace is obviously aware of the gc.
Is the only way I saw to have gc's that are inline between Ops and Opcodes